### PR TITLE
8337876: [IR Framework] Add support for IR tests with @Stable

### PR DESCRIPTION
**Pull request by chhagedorn**
It is currently not possible to write IR tests with `@Stable` annotations because one need to somehow add the IR test classes to the boot classpath. This patch provides support to write such IR tests. I've added a section to the README to provide guidance on how this can be done.

This is motivated by https://github.com/openjdk/jdk/pull/19635.

I've tested this patch by taking the current patch of https://github.com/openjdk/jdk/pull/19635, dropping the `RestrictStable` flag and modifying the tests to work with the new IR framework feature:
```
TestFramework testFramework = new TestFramework();
testFramework
        .addFlags("-XX:-TieredCompilation",
                  "-XX:+UseParallelGC")
        .addTestClassesToBootClassPath()
        .start();
```

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337876](https://bugs.openjdk.org/browse/JDK-8337876): [IR Framework] Add support for IR tests with @Stable (**Enhancement** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20477/head:pull/20477` \
`$ git checkout pull/20477`

Update a local copy of the PR: \
`$ git checkout pull/20477` \
`$ git pull https://git.openjdk.org/jdk.git pull/20477/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20477`

View PR using the GUI difftool: \
`$ git pr show -t 20477`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20477.diff">https://git.openjdk.org/jdk/pull/20477.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20477#issuecomment-2271105119)

### CODE REVIEW

**Comment by chhagedorn** [comment]
@shipilev @liach Since you've already looked at this idea in https://github.com/openjdk/jdk/pull/19635, I'm kindly asking you to also have a look at this PR, thanks!

File: test/hotspot/jtreg/compiler/lib/ir_framework/README.md [hunk 1]
```diff
@@ -157,7 +157,16 @@ The framework allows the use of additional compiler control annotations for help
 - [@ForceCompile](./ForceCompile.java)
 - [@ForceCompileClassInitializer](./ForceCompileClassInitializer.java)
 
-### 2.5 Framework Debug and Stress Flags
+### 2.5 IR Tests with `@Stable` annotation
+To run tests with `@Stable` annotations, one need to add the test classes to the boot classpath. This can easily be achieved by calling `TestFramework.addTestClassesToBootClassPath()` on the test framework object:
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/lib/ir_framework/README.md:161]: I think mentioning `@Stable` here is overly specific. This guidance applies to any code that expects to be run in privileged mode, whether it is `@Stable`, `@Contended`, `@ReservedStackAccess`, etc. I say it should be e.g. "2.5 IR Tests With Privileged Classes" or some such.

File: test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java [hunk 2]
```diff
@@ -323,6 +324,15 @@ public TestFramework addScenarios(Scenario... scenarios) {
         return this;
     }
 
+    /**
+     * Add test classes to boot classpath. This adds all classes found on path {@link jdk.test.lib.Utils#TEST_CLASSES}
+     * to the boot classpath with "-Xbootclasspath/a". This is useful when trying to run tests with @Stable annotations.
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java:329]: Again, `@Stable` is overly specific here. This is "just" about the privileged code.

File: test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java [hunk 3]
```diff
@@ -91,15 +91,20 @@ public static String getLastTestVMOutput() {
...
     private void prepareTestVMFlags(List<String> additionalFlags, TestFrameworkSocket socket, Class<?> testClass,
-                                    Set<Class<?>> helperClasses, int defaultWarmup) {
+                                    Set<Class<?>> helperClasses, int defaultWarmup, boolean testClassesOnBootClassPath) {
         // Set java.library.path so JNI tests which rely on jtreg nativepath setting work
         cmds.add("-Djava.library.path=" + Utils.TEST_NATIVE_PATH);
         // Need White Box access in test VM.
-        cmds.add("-Xbootclasspath/a:.");
+        String bootClassPath = "-Xbootclasspath/a:.";
+        if (testClassesOnBootClassPath) {
+            // Add test classes themselves to boot classpath. This is required, for example, for IR tests with @Stable.
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java:100]: Here too, drop the mention of `@Stable`?

File: test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java [hunk 4]
```diff
@@ -91,15 +91,20 @@ public static String getLastTestVMOutput() {
...
+            bootClassPath += ":" + Utils.TEST_CLASSES;
+        }
+        cmds.add(bootClassPath);
         cmds.add("-XX:+UnlockDiagnosticVMOptions");
         cmds.add("-XX:+WhiteBoxAPI");
-        // Ignore CompileCommand flags which have an impact on the profiling information.
-        List<String> jtregVMFlags = Arrays.stream(Utils.getTestJavaOpts()).filter(s -> !s.contains("CompileThreshold")).collect(Collectors.toList());
+        // Ignore CompileThreshold and CompileCommand flags which have an impact on the profiling information.
+        List<String> jtregVMFlags = Arrays.stream(Utils.getTestJavaOpts()).filter(s -> !s.contains("CompileThreshold")).toList();
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java:107]: This hunk seems unnecessary? I would like to have IR Framework patch that is easily backportable and does not contain additional, non-essential hunks :)

**Review by shipilev — COMMENTED** [review]
Looks okay, I just think we don't need to mention `@Stable`.

File: test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java (hunk 4 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java:107]: Fair point, could be done separately, dropped :)

File: test/hotspot/jtreg/compiler/lib/ir_framework/README.md (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/lib/ir_framework/README.md:161]: Thanks for the quick review! You're right, that's too specific. I've pushed an update according to your suggestion.

File: test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java [hunk 5]
```diff
@@ -91,11 +92,16 @@ public static String getLastTestVMOutput() {
...
-                                    Set<Class<?>> helperClasses, int defaultWarmup) {
+                                    Set<Class<?>> helperClasses, int defaultWarmup, boolean testClassesOnBootClassPath) {
         // Set java.library.path so JNI tests which rely on jtreg nativepath setting work
         cmds.add("-Djava.library.path=" + Utils.TEST_NATIVE_PATH);
         // Need White Box access in test VM.
-        cmds.add("-Xbootclasspath/a:.");
+        String bootClassPath = "-Xbootclasspath/a:.";
+        if (testClassesOnBootClassPath) {
+            // Add test classes themselves to boot classpath to make them privileged.
+            bootClassPath += ":" + Utils.TEST_CLASSES;
```
[the remark this answers is not in the document]
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/lib/ir_framework/driver/TestVMProcess.java]: Good catch! Fixed.

**Review by chhagedorn — COMMENTED** [review]
Sorry for the delay. That is definitely a good idea to add a simple test to check the new feature. I've added such a test with `@Stable` that fails when not adding the test class to the boot classpath and works otherwise.

**Review by shipilev — APPROVED** [review]
OK, thanks.

**Comment by chhagedorn** [comment]
Thanks Aleksey for your review!

**Comment by shipilev** [comment]
Can/should we get some more reviews here? I would like to get the `@Stable` PR, which depends on it, moving :)

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

/integrate

**Pull request closed without merging**